pkg/providers/tinkerbell/hardware/validator.go

```diff
@@ -231,3 +232,67 @@ func validateLabelValue(v string) error {
 	}
 	return nil
 }
+
+// MachineSelector is a map of key-value pairs used to select Machine instances by label definition.
+type MachineSelector map[string]string
+
+func (m MachineSelector) String() string {
+	s, err := json.Marshal(m)
+	// We want to support fmt.Stringer so we have no choice but to panic here. In practice, this
+	// should never happen because we're merely stringifying a map.
+	if err != nil {
+		panic(err)
```

Comment:
shouldn't panic here, use Errorf

Reply:
What do we do with the error returned from `fmt.Errorf()`?

Reply:
We have generally tried to avoid using panic. Not sure if it is spelled out in the code guidelines but panic should be used sparingly and reserved for programmer errors. I agree that it could be a gray area here, but bubbling an error back up has been the general pattern and would be good for consistency and to avoid blurring boundaries and abusing it by more casual use